tests/raft_scenarios_runner.py

```diff
@@ -36,18 +36,45 @@ def write_error_report(errors=None):
         print("??? success \n")
 
 
-def remove_reconfiguration_replicates(log):
+def preprocess_for_trace_validation(log):
+    """
+    For each node, remove the last "replicate" or "execute_append_entries_sync"
+    preceding an "add_configuration" entry. For the first node, check the initial
+    expected sequence of "become_leader", "add_configuration", "replicate" (committable),
+    followed by "commit", and replace it with a "bootstrap" entry.
+    """
+    # Log may be empty if CCF_RAFT_TRACING=OFF
+    if not log:
+        return log
     log_by_node = defaultdict(list)
+    initial_node = None
     for line in log:
         entry = json.loads(line)
         node = entry["msg"]["state"]["node_id"]
+        if initial_node is None:
+            initial_node = node
         if entry["msg"]["function"] == "add_configuration":
             removed = log_by_node[node].pop()
             assert removed["msg"]["function"] in (
                 "replicate",
                 "execute_append_entries_sync",
             ), removed
         log_by_node[node].append(entry)
+
+    def head():
+        return log_by_node[initial_node].pop(0)
+
+    assert head()["msg"]["function"] == "become_leader"
+    assert head()["msg"]["function"] == "add_configuration"
+    signature = head()
+    assert signature["msg"]["function"] == "replicate", signature
+    assert signature["msg"]["globally_committable"], signature
+    commit = head()
+    assert commit["msg"]["function"] == "commit", commit
+    assert commit["msg"]["state"]["commit_idx"] == 2, commit
+    # Commit becomes bootstrap, the entry point into the trace validation
+    commit["msg"]["function"] = "bootstrap"
```

Comment:
Could `raft.h` log it under "bootstrap" depending on its internal state, instead of interfering here?

Reply:
Perhaps, but not without extra state, and interfering with `raft.h` is much more risky and unappealing than massaging the trace a bit. My mental model is that raft traces should be function calls, with a pre-state and relevant arguments (*). This not 100% true just now, but we're not far off.

Pushing complexity into the implementation because of limitations of the trace validation does not seem justifiable to me. In a perfect world we would unwind the bootstrap and make the spec able to cope with the full details of startup instead.

Also we may want to have a more generic version of this in the future for #5290, which could bootstrap from arbitrary states on any given trace. It's much easier to experiment with this in a script that in the production code.

(*) One because it's then very easy to ifdef them out and convince ourselves they have 0 runtime impact in release builds, and two because it means we probably never need to make changes to them, which means that we diminish the chance that we will ever impact release builds because of them.